symbolic-unwind/src/evaluator/mod.rs

```diff
@@ -451,6 +452,130 @@ pub enum Identifier {
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Rule<A: RegisterValue>(Identifier, Expr<A>);
 
+/// A half-open range `[start, end)` with some `contents` attached.
+#[derive(Clone, Debug)]
+pub struct Range<A, E> {
+    /// The left endpoint of the range.
+    pub start: A,
+
+    /// The right endpoint of teh range.
+    pub end: A,
+
+    /// The contents associated with the range.
+    pub contents: E,
+}
+
+impl<A: Eq, E> PartialEq for Range<A, E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.end == other.end && self.start == other.start
+    }
+}
+
+impl<A: Eq, E> Eq for Range<A, E> {}
+
+impl<A: Ord, E> core::cmp::Ord for Range<A, E> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.end.cmp(&other.end).then(self.start.cmp(&other.start))
+    }
+}
+
+impl<A: Ord, E> PartialOrd for Range<A, E> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A structure containing a set of disjoint half-open ranges `[a, b)` with attached contents.
+#[derive(Clone, Debug)]
+pub struct RangeMap<A, E> {
+    inner: Vec<Range<A, E>>,
+}
+
+impl<A: Ord + Copy, E> RangeMap<A, E> {
+    /// Insert a range into the map.
+    ///
+    /// The range must be disjoint from all ranges that are already present.
+    /// Returns true if the insertion was successful.
+    pub fn insert(&mut self, start: A, end: A, contents: E) -> bool {
```

Comment:
can this also implement `FromIterator`?  It would be nice to be able to create this using `Iterator::collect` which is somewhat how the equivalent thing in symbolicator currently is constructed.

Reply:
Well, that depends. Can `FromIterator` be fallible? Because not every iterator of Ranges would result in a valid `RangeMap`. Otherwise I suppose we could have later ranges overwrite earlier ones in case they conflict.

Reply:
good point, how about a manual `from_iter()` that returns a `Result`?

Reply:
well, `try_from_iter()`, `from_iter()` is already taken by `FromIterator` :)

Reply:
Yeah, that should be easy enough. But a custom method wouldn't allow one to use `collect`, right? Should we have a `from_iter` with "reasonable" non-fallible behavior, like just ignoring ranges that can't be inserted?

Reply:
hmm, i think it's better to forgo `.collect()` instead of silently ignoring errors.  It's not what you expect from this API to a more explicit version that is fallible is better.